pymc3/distributions/multivariate.py

```diff
@@ -690,6 +691,138 @@ def logp(self, x):
+        of the second dimension of *a*.
+
+    shape : numerical tuple
+        Describes shape of distribution. For example if n=array([5, 10]), and
+        p=array([1, 1, 1]), shape should be (2, 3).
+    """
+
+    def __init__(self, n, a, shape, *args, **kwargs):
+        super().__init__(shape, *args, **kwargs)
+
+        if len(self.shape) > 1:
+            self.n = tt.shape_padright(n)
+            self.a = tt.as_tensor_variable(a) if np.ndim(a) > 1 else tt.shape_padleft(a)
+        else:
+            # n is a scalar, p is a 1d array
+            self.n = tt.as_tensor_variable(n)
+            self.a = tt.as_tensor_variable(a)
+
+        p = self.a / self.a.sum(-1, keepdims=True)
+
+        self.mean = self.n * p
+        mode = tt.cast(tt.round(self.mean), "int32")
+        diff = self.n - tt.sum(mode, axis=-1, keepdims=True)
+        inc_bool_arr = tt.abs_(diff) > 0
+        mode = tt.inc_subtensor(mode[inc_bool_arr.nonzero()], diff[inc_bool_arr.nonzero()])
+        self.mode = mode
+
+    def _random(self, n, a, size=None, raw_size=None):
+        # numpy will cast dirichlet and multinomial samples to float64 by default
+        original_dtype = a.dtype
+
+        # Thanks to the default shape handling done in generate_values, the last
+        # axis of n is a dummy axis that allows it to broadcast well with `a`
+        n = np.broadcast_to(n, size)
+        a = np.broadcast_to(a, size)
+        n = n[..., 0]
+
+        # np.random.multinomial needs `n` to be a scalar int and `a` a
+        # sequence so we semi flatten them and iterate over them
+        n_ = n.reshape([-1])
+        a_ = a.reshape([-1, a.shape[-1]])
+        p_ = np.array([np.random.dirichlet(aa) for aa in a_])
+        samples = np.array([np.random.multinomial(nn, pp) for nn, pp in zip(n_, p_)])
+        samples = samples.reshape(a.shape)
+
+        # We cast back to the original dtype
+        return samples.astype(original_dtype)
+
+    def random(self, point=None, size=None):
+        """
+        Draw random values from Dirichlet-Multinomial distribution.
+
+        Parameters
+        ----------
+        point: dict, optional
+            Dict of variable values on which random values are to be
+            conditioned (uses default point if not specified).
+        size: int, optional
+            Desired size of random sample (returns one sample if not
+            specified).
+
+        Returns
+        -------
+        array
+        """
+        n, a = draw_values([self.n, self.a], point=point, size=size)
+        samples = generate_samples(
+            self._random,
+            n,
+            a,
+            dist_shape=self.shape,
+            not_broadcast_kwargs={"raw_size": size},
+            size=size,
+        )
+
+        if size is not None:
+            expect_shape = (size, *self.shape)
+        else:
+            expect_shape = self.shape
+        assert tuple(samples.shape) == tuple(expect_shape)
```

Comment:
Do we want this here? Is any other distribution doing the same?

Reply:
I feel like it is unnecessary during runtime, since we are already testing this quite a lot in the unittests

Reply:
Asserts are not evil, but also not necessarily the best option. (Also see https://stackoverflow.com/a/13534633)

To validate external inputs, use
```
if input_shape != expected_shape:
    raise ShapeError("Your fault!")
```
This has the advantage that a coverage check can tell you if the tests cover the exception.

If you want to validate an internal assumption and make sure that your code did not mess up, the assert is the right thing to do. It can help tremendously to debug or understand code.

In https://github.com/michaelosthege/pyrff/blob/master/pyrff/rff.py#L230-L261 I did both, because it took me weeks to understand the shapes of the code I was re-implementing there...


In this case if you're already testing it a lot maybe a comment instead of an assert is more appropriate.

Reply:
> In this case if you're already testing it a lot maybe a comment instead of an assert is more appropriate.

Yes, I think this is being sufficiently tested for reasonable inputs; I believe @ricardoV94 demonstrated already that _these_ asserts were passed for a large variety of shapes (...even when some of the actual outputs were somewhat unintuitive).

I'm more worried about users "abusing" weirdly shaped inputs, in which case I like your explicit ShapeError to catch corner cases we didn't even think of.

On the other hand, Multinomial doesn't do this now. If we wanted to add this check I think we should do it for both distributions in parallel.

Reply:
> On the other hand, Multinomial doesn't do this now. If we wanted to add this check I think we should do it for both distributions in parallel.

I agree with this